[Uptime][Docs] Add deployment architecture section (#43028)
elastic/kibana: docs/uptime-guide/deployment-arch.asciidoc

[role="xpack"]
[[uptime-deployment-arch]]
== Deployment Architecture

There are infinite ways to deploy Uptime and Heartbeat, some of which work better than others. A guiding principle in these recommendations is that an outage that takes down the service being monitored should not also take down Heartbeat. You will want Heartbeat to be functioning even when your site is not, so the guidelines that follow attempt to minimize this possibility.

> Heartbeat is generally run as a centralized service within a datacenter. While it is possible to run it as a separate "sidecar" process paired with each process/container we recommend against it. Running Heartbeat centrally ensures you will still be able to see monitoring data in the event of an overloaded, disconnected, or otherwise malfunctioning server. 

Reviewer: Consistency, and rewording slightly to clearly state up-front that we do **not** recommend something. 

```suggestion
Heartbeat is generally run as a centralized service within a data center. While it is possible to run it as a separate "sidecar" process paired with each process or container, we do not recommend this. Running Heartbeat centrally ensures you will still be able to see monitoring data in the event of an overloaded, disconnected, or otherwise malfunctioning server. 
```

> For further redundancy, you may want to deploy multiple Heartbeats across geographic and/or network boundaries to provide more data. Specify Heartbeat's observer https://www.elastic.co/guide/en/beats/heartbeat/current/configuration-observer-options.html[geo options] to do so. Some examples might be:

Reviewer: When you fix the link, can you check the exact wording in the area of the link? As written, I think it's going to read "Specify Heartbeat's observer geo options" where "geo options" is the link. The heading on the linked page is "Specify observer and geo options". If you want to refer the reader to just the geo options, then you should probably drop "observer" in the text here. If it's both, then the linked text should probably be "observer and geo options".

> For further redundancy, you may want to deploy multiple Heartbeats across geographic and/or network boundaries to provide more data. Specify Heartbeat's observer https://www.elastic.co/guide/en/beats/heartbeat/current/configuration-observer-options.html[geo options] to do so. Some examples might be:

Author: I've seen us use `{beatname} agent` and just plain `{beatname}` in various places. Heartbeat isn't really an agent (an agent implies it being a sidecar, which heartbeat is not).

My sense here is that colloquially the `agent/instance` clarification is generally implied by context and omitted for brevity, and only explicitly added for additional emphasis and disambiguation where required.

Would love to hear from @dedemorton for some guidance here.

> For further redundancy, you may want to deploy multiple Heartbeats across geographic and/or network boundaries to provide more data. Specify Heartbeat's observer https://www.elastic.co/guide/en/beats/heartbeat/current/configuration-observer-options.html[geo options] to do so. Some examples might be:

Reviewer 2: @andrewvc We shouldn't say Heartbeats (plural). Most style guides advise against making product names plural because it impacts branding and trademarking. 

My recommendation is:

- When you only have one instance of a Beat running, there's no need to qualify it.

- If you are talking about multiple instances, you should say "Heartbeat instances" rather than "Heartbeats".

Also, you should use the `{heartbeat}` shared attribute to resolve the Beat name. This makes it easier for us to change the name if we ever have to go through rebranding. Using the attribute also safeguards against weird typos, like Heatbeat. hehe.

> * **A site served from a CDN with points of presence (POPs) around the globe:** In this case you may want to have multiple heartbeats at different data-centers around the world checking to see if your site is reachable via local CDN points of presence.

Reviewer: Will our readers be likely to know what a CDN is? Yet be unlikely to know what PoP is? Would this be more consistent as below? 

Also note I've gone for "Heartbeat instances" here, but I'm open to other options, so long as it's consistent.
```suggestion
* **A site served from a content delivery network (CDN) with points of presence (POPs) around the globe:** In this case you may want to have multiple Heartbeat instances at different data centers around the world checking to see if your site is reachable via local CDN POPs.
```

> * **A service within a single data center that is accessed across multiple VPNs:** Setup one heartbeat instance within the VPN the service operates from, and another within an additional VPN users access the server from. Having both will help pinpoint network errors in the event of an outage.

Reviewer: Here, I'm suggesting a couple of minor consistency changes, and an added "that" to help disentangle the adjectives and nouns in the sentence. Otherwise it may be read as "additional VPN users", whereas it's presumably "additional VPNs"?

```suggestion
* **A service within a single data center that is accessed across multiple VPNs:** Set up one Heartbeat instance within the VPN the service operates from, and another within an additional VPN that users access the service from. Having both instances will help pinpoint network errors in the event of an outage.
```

> * **A single service running primarily in a datacenter on the US east coast, with a hot failover located in a US west coast datacenter:** Run a Heartbeat instance in each datacenter checking both its local copy of the service and its counterpart across the country. You would have two monitors in each region, one for the local service, one for the remote. In the event of a datacenter failure it would be immediately obvious if the service had a connectivity issue to the outside world or if the failure was only internal.

Reviewer: A couple more consistency changes, and a bit more clarifying and disentangling. Sorry, but for me, this comment isn't showing the differences between the existing and suggested text. I hope you have a better view!

```suggestion
* **A single service running primarily in a US east coast data center, with a hot failover located in a US west coast datacenter:** In each data centre, run a Heartbeat instance that checks both the local copy of the service and its counterpart across the country. You would have two monitors in each region, one for the local service and one for the remote service. In the event of a data center failure it will be immediately obvious if the service had a connectivity issue to the outside world or if the failure was only internal.
```
